mds: remove subtree map from journal

ukernel opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

ukernel committed:
mds: track subtree bounds in export/import state

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: unify uncommitted leader/peers/fragments waiting

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: index export state by dirfrag_t

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: use standard two-phrase commit protocol to resolve ambiguous import

Later patch will make mds not journal subtrees they own. So importer and
bystanders can no longer use 'if a subtree is claimed by exporter' to
disambiguate subtree's auth.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: avoid adding non-auth metadata to journal

This implies that, after replaying journal, mds' cache may have
multiple disconnected subtrees.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: only add root/mdsdir and ambiguous imports to ESubtreeMap

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: reconnect subtrees during cache rejoin

The basic idea is mds share cached subtrees auth with other recovering
mds. During reconnecting subtrees, the recovering mds checks learned
subtrees auth to decide if they should open or discover given dirfrags.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
include/frag: dedup overlapping entries in fragset_t

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: reconnect caps and snaprealms after subtrees are connected

unable to setup caps' parent snaprealm and snaprealm's parent when
subtrees are disconnected.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: add incompat feature "no subtreemap"

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: update projected version after rollback dir fragments

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: simplify MDCache::adjust_bounded_subtree_auth()

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel commented:
jenkins test make check

batrick commented:
jenkins test api

ukernel committed:
mds: fast export subtrees that are empty and clean

To export a clean subtree, exportor just needs to submit an EExport to
log. If exporter fails, the EExport ensures that the subtree is not in
exporter's cache after log replay. The importer just needs to add the
subtree to its cache.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: unify trimming subtree and non-subtree dirfrags

Only PIN_SUBTREE for base subtree and auth subtree's non-auth bounds.
For auth subtree, don't PIN_REPLICATED when there is only one replica.
Because the subtree is always replicated to auth mds of subtree's
parent inode.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel committed:
mds: don't create aux subtree for dir pin

replace aux subtree with aux bound flag. freeze_tree() stops walking
down when it encounters dirfrags with the aux bound flag.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

ukernel commented:
jenkins test make check

ukernel commented:
jenkins retest this please

ukernel committed:
mds: clear CInode::mds_caps_wanted if import aborts

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

batrick commented:
@ukernel just FYI, we're planning to defer merging this to mid-January after pacific is branched off. This changeset is a bit too significant to merge now.

vshankar reviewed (commented):
Need to go through the changes with more detail. Some comments/suggestions to start with.

Commit:
mds: make MDCache::adjust_dir_fragments() handle undef dirfrags

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: introduce MDSCacheObject::is_rejoin_undef()

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: send strong cache rejoin after target's subtrees are connected

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

ukernel commented on src/mds/LogSegment.h:
```diff
@@ -88,10 +88,16 @@ class LogSegment {
-  set<metareqid_t> uncommitted_leaders;
-  set<metareqid_t> uncommitted_peers;
+
+  set<metareqid_t> uncommitted_leaders_peers;
```
it's cleanup

ukernel commented on src/mds/Migrator.cc:
```diff
@@ -1360,12 +1347,13 @@ void Migrator::export_frozen(CDir *dir, uint64_t tid)
 
   // generate prep message, log entry.
-  auto prep = make_message<MExportDirPrep>(dir->dirfrag(), it->second.tid);
+  auto prep = make_message<MExportDirPrep>(dir->dirfrag(), it->second.bound_vec, it->second.tid);
```
> Follow up question (for my understanding): Is a subtree "bound" a collection of all subtree roots? (where each subtree root could either be a CDir* or a dirfrag_t).

right

ukernel replied:
> The journal entry now has a vector of dirfrags. This used to be just subtree dirfrag and tid earlier. This would increase the size of the journal entry. Would that have other side-effects?

I don't think so

Commit:
mds: make Locker::eval() check snaplock

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: check authpinnability in Locker::rdlock_try_set()

Otherwise, Locker::rdlock_try_set() may change frozen object's lock
state to unstable state.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: make Locker::wrlock_force() handle coexisting remote wrlock

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: handle race between fragment notify and cache expire

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't skip dirty null dentries when committing fragmenting dirfrag

CDir::is_new() can be false postive (corresponding dirfrag had already
committed to object store) after journal replay.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: introduce Session::is_open_or_stale()

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't drop replayed request when session is stale

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: call MDSRank::queue_one_replay() after dropping replayed request

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: properly check if peer request needs rollback

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: keep filelock of straydirs writeable as long as possible

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

ukernel committed:
mds: fix leaking frozen inode during rename rollback

mdr->more()->is_ambiguous_auth can be false when rename_inode is frozen.

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

Commit:
mds: consider 'in' mds when hashing inode/dirfrag into rank

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

ukernel committed:
mds: avoid concurrently rename and fragment directory

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

Commit:
mds: simplify deadlock avoidance of freezing tree/dir

Do not check if there are remote authpin waiters. Just cancel
export/fragment if tree/dir does not become frozen in time.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: remove wrong debug assertion

Unlink operation with lock cache can fail to auth pin straydir if
straydir is being fragmented.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't set xlock of snaplock done when early replying rename

there are code that use either dentry lock or snaplocks to lock path.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't send message to mds that are in replay state

The mds can still be in standby state (it hasn't got mdsmap) if we
send message to it. So connection to the mds can have wrong peer
information.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: resync straydir's fragstat/rstat when purging stray

After fragmenting, straydir's fragstat/rstat can be stale

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: ensure all peers get cache rejoin acks before going to active

Otherwise peer rename request can update replicated srcdn in bystander
mds, then the bystander mds gets cache rejoin ack from srcdn's auth mds.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: set mdr->more()->is_inode_exporter iif after exporting inode

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't evict importing session when revoking stale write caps

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't send pending rejoins if cluster is no long in rejoining state

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: add perf counters for import/export failures

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: avoid assertion after resetting sessionmap

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: avoid passing base inode number to MDCache::open_ino()

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't hold wrlock while requesting remote wrlock

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: fix nullptr dereference when rolling back peer rename

CInode is still linked to srcdn if it's still freezing inode

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: don't early reply if subtree auth is unstable for failover

if a subtree is not durable in journal, its auth may change after
mds failover.

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

Commit:
mds: revoke async dirop caps if subtree auth is unstable for failover

Signed-off-by: "Yan, Zheng" <yanzheng03@kuaishou.com>

ukernel commented:
ping @batrick

batrick commented:
Will run this through testing now and review in the next week.

batrick commented:
https://pulpito.ceph.com/pdonnell-2021-03-25_22:37:32-fs-wip-pdonnell-testing-20210325.205235-distro-basic-smithi/

batrick commented:
rebased to fix options.cc conflict

vshankar commented:
@ukernel Given the complexity of this feature and the PR itself, it would really help to have a design overview and the approach detailed out as a doc. BTW, I have this in my review backlog for later this week (and next week).

orange30 commented:
@ukernel  
I add a new bug report:
https://tracker.ceph.com/issues/57920

Will you try to push this commit again?
